File: znai-reactjs/src/doc-elements/api/ApiParameter.js
/*
 * Copyright 2020 znai maintainers
 * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react'

import ApiParameters from './ApiParameters'
import {ReferenceLinkWrapper} from '../references/ReferenceLinkWrapper'

import {Icon} from '../icons/Icon';
import './ApiParameter.css'

class ApiParameter extends React.Component {
    constructor(props) {
        super(props)

        this.state = { isExpanded: this.props.isExpanded, width: 0 }
    }

    render() {
        const {
            anchorId,
            name,
            type,
            children,
            description,
            nestedLevel = 0,
            elementsLibrary
        } = this.props

        const {
            isExpanded,
            width
        } = this.state

        const reference = this.findReference()

        const commonClassName = ' znai-api-param-cell' +
            (isExpanded ? ' expanded' : '') +
            (children ? ' expandable' : '') +
            (reference ? ' with-reference' : '') +
            (nestedLevel > 0 ? ' nested-' + nestedLevel : '')

        const nameTypeClassName = 'znai-api-param-name-type-toggle-cell' + commonClassName
        const descriptionClassName = 'znai-api-param-description-cell' + commonClassName

        const toggleOnClick = children ? this.toggleExpand : null

        const expandToggle = children && (
            <div className="expand-toggle">
                {isExpanded ? '-' : '+'}
            </div>)

        const renderedChildren = (children && isExpanded) ? (
            <React.Fragment>
                <div/>
                <ApiParameters parameters={children}
                               nestedLevel={nestedLevel + 1}
                               parentWidth={width}
                               elementsLibrary={elementsLibrary}/>
            </React.Fragment>
        ): null

        const referenceUrl = reference ? reference.pageUrl : null

        return (
            <React.Fragment>
                <div className={nameTypeClassName}>
                    <div className="znai-api-param-name-type-toggle" ref={node => this.nameAndTypeNode = node}>
                        <div className="znai-api-param-name">
                            <ReferenceLinkWrapper referenceUrl={referenceUrl}>
                                {name}
                            </ReferenceLinkWrapper>
                        </div>
                        <div className="znai-api-param-type-and-toggle" onClick={toggleOnClick}>
                            {expandToggle}
                            <div className="znai-api-param-type">
                                {type}
                            </div>
                        </div>
                    </div>
                    <div className="znai-api-param-anchor" id={anchorId}>
                        {anchorId && <a href={"#" + anchorId}><Icon id="link"/></a>}
                    </div>
                </div>
                <div className={descriptionClassName}>
                    <elementsLibrary.DocElement content={description} elementsLibrary={elementsLibrary}/>
                </div>
                {renderedChildren}
            </React.Fragment>
        )
    }

    componentDidMount() {
        const width = this.nameAndTypeNode.getBoundingClientRect().width
        this.setState({width})
    }

    toggleExpand = () => {
        this.setState(prev => ({isExpanded: !prev.isExpanded}))
    }

    findReference() {
        const {references, type, name} = this.props
        if (!references) {
            return null
        }

        const byType = references[type]
        if (byType) {
            return byType
        }

        const byName = references[name]
        if (byName) {
            return byName
        }
    }
}

export default ApiParameter

File: znai-core/src/main/java/com/twosigma/znai/extensions/api/ApiParameter.java
/*
 * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.twosigma.znai.extensions.api;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.toList;

public class ApiParameter {
    private final String name;
    private final String type;
    private final List<Map<String, Object>> description;

    private final List<ApiParameter> children;

    public ApiParameter(String name, String type, List<Map<String, Object>> description) {
        this.name = name;
        this.type = type;
        this.description = description;
        this.children = new ArrayList<>();
    }

    public String getName() {
        return name;
    }

    public String getType() {
        return type;
    }

    public List<Map<String, Object>> getDescription() {
        return description;
    }

    public List<ApiParameter> getChildren() {
        return children;
    }

    public ApiParameter add(String name, String type, List<Map<String, Object>> description) {
        ApiParameter apiParameter = new ApiParameter(name, type, description);
        children.add(apiParameter);

        return apiParameter;
    }

    ApiParameter find(String name) {
        return children.stream()
                .filter(p -> p.name.equals(name))
                .findFirst()
                .orElseThrow(() -> new RuntimeException("no children found with name: '" + name +
                        "' in '" + this.name + "'"));
    }

    public Map<String, Object> toMap() {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("name", name);
        result.put("type", type);
        result.put("description", description);
        if (! children.isEmpty()) {
            result.put("children", children.stream().map(ApiParameter::toMap).collect(toList()));
        }

        return result;
    }
}
